File: server/src/plugins/runtime/node/storageController.ts
import { DeviceStorage } from './storage.js';

import type { RPCClient } from '@camera.ui/rpc';
import type { JsonSchema, PluginInfo, SensorType } from '@camera.ui/sdk';
import type { PluginConfigDb } from './configDb.js';
import type { PluginAPI } from './pluginApi.js';

// Stable storage key for sensor instance storage.
function sensorStorageKey(cameraId: string, sensorType: SensorType, pluginId: string, sensorName: string): string {
  return `${cameraId}:sensor:${sensorType}:${pluginId}:${sensorName}`;
}

export class StorageController {
  #api: PluginAPI;
  #proxy: RPCClient;
  #plugin: PluginInfo;
  #pluginDb: PluginConfigDb;
  #storages = new Map<string, DeviceStorage<any>>();

  constructor(api: PluginAPI, proxy: RPCClient, plugin: PluginInfo, pluginDb: PluginConfigDb) {
    this.#api = api;
    this.#proxy = proxy;
    this.#plugin = plugin;
    this.#pluginDb = pluginDb;
  }

  public createCameraStorage<T extends Record<string, any> = Record<string, any>>(cameraId: string, schemas: JsonSchema[] = []): DeviceStorage<T> {
    let cameraStorage = this.#storages.get(cameraId);

    if (!cameraStorage) {
      cameraStorage = new DeviceStorage<T>(this.#api, this.#proxy, this.#plugin, this.#pluginDb, { kind: 'camera', cameraId }, schemas);
      this.#storages.set(cameraId, cameraStorage);
    } else {
      cameraStorage.updateSchema(schemas);
    }

    return cameraStorage as DeviceStorage<T>;
  }

  public createPluginStorage<T extends Record<string, any> = Record<string, any>>(schemas: JsonSchema[] = []): DeviceStorage<T> {
    let pluginStorage = this.#storages.get('storage');

    if (!pluginStorage) {
      pluginStorage = new DeviceStorage<T>(this.#api, this.#proxy, this.#plugin, this.#pluginDb, { kind: 'plugin' }, schemas);
      this.#storages.set('storage', pluginStorage);
    } else {
      pluginStorage.updateSchema(schemas);
    }

    return pluginStorage as DeviceStorage<T>;
  }

  public getCameraStorage<T extends Record<string, any> = Record<string, any>>(deviceId: string): DeviceStorage<T> | undefined {
    return this.#storages.get(deviceId) as DeviceStorage<T> | undefined;
  }

  public getPluginStorage<T extends Record<string, any> = Record<string, any>>(): DeviceStorage<T> | undefined {
    return this.#storages.get('storage') as DeviceStorage<T> | undefined;
  }

  public createSensorStorage<T extends Record<string, any> = Record<string, any>>(
    cameraId: string,
    sensorType: SensorType,
    pluginId: string,
    sensorName: string,
    sensorId: string,
    schemas: JsonSchema[] = [],
  ): DeviceStorage<T> {
    const storageKey = sensorStorageKey(cameraId, sensorType, pluginId, sensorName);
    let storage = this.#storages.get(storageKey);

    if (!storage) {
      // storageKey is only the in-memory registry key; persistence addresses
      // the canonical sensors.<camId>.<type>.<name> path. sensorId is the
      // runtime UUID for the RPC namespace.
      storage = new DeviceStorage<T>(this.#api, this.#proxy, this.#plugin, this.#pluginDb, { kind: 'sensor', cameraId, sensorType, sensorName }, schemas, sensorId);
      this.#storages.set(storageKey, storage);
      storage.updateSchema(schemas);
    } else {
      storage.updateSchema(schemas);
    }

    return storage as DeviceStorage<T>;
  }

  public getSensorStorage<T extends Record<string, any> = Record<string, any>>(
    cameraId: string,
    sensorType: SensorType,
    pluginId: string,
    sensorName: string,
  ): DeviceStorage<T> | undefined {
    const storageKey = sensorStorageKey(cameraId, sensorType, pluginId, sensorName);
    return this.#storages.get(storageKey) as DeviceStorage<T> | undefined;
  }

  public async createStorage(type: 'camera', deviceId: string): Promise<DeviceStorage<any>>;
  public async createStorage(type: 'plugin'): Promise<DeviceStorage<any>>;
  public async createStorage(
    type: 'sensor',
    cameraId: string,
    sensorType: SensorType,
    pluginId: string,
    sensorName: string,
    sensorId: string,
  ): Promise<DeviceStorage<any>>;
  public async createStorage(
    type: 'camera' | 'plugin' | 'sensor',
    deviceIdOrCameraId?: string,
    sensorTypeOrUndefined?: SensorType,
    pluginId?: string,
    sensorName?: string,
    sensorId?: string,
  ): Promise<DeviceStorage<any>> {
    let storage: DeviceStorage<any> | undefined;

    if (type === 'camera') {
      if (!deviceIdOrCameraId) {
        throw new Error('deviceId is required for camera storage creation');
      }
      storage = this.createCameraStorage(deviceIdOrCameraId);
    } else if (type === 'sensor') {
      if (!deviceIdOrCameraId || !sensorTypeOrUndefined || !pluginId || !sensorName || !sensorId) {
        throw new Error('cameraId, sensorType, pluginId, sensorName and sensorId are required for sensor storage creation');
      }
      storage = this.createSensorStorage(deviceIdOrCameraId, sensorTypeOrUndefined, pluginId, sensorName, sensorId);
    } else {
      storage = this.createPluginStorage();
    }

    await storage.registerStorage();
    return storage;
  }

  public async removeStorage(type: 'camera', deviceId: string): Promise<void>;
  public async removeStorage(type: 'plugin'): Promise<void>;
  public async removeStorage(type: 'sensor', cameraId: string, sensorType: SensorType, pluginId: string, sensorName: string): Promise<void>;
  public async removeStorage(
    type: 'camera' | 'plugin' | 'sensor',
    deviceIdOrCameraId?: string,
    sensorTypeOrUndefined?: SensorType,
    pluginId?: string,
    sensorName?: string,
  ): Promise<void> {
    let storageKey: string;

    if (type === 'sensor') {
      if (!deviceIdOrCameraId || !sensorTypeOrUndefined || !pluginId || !sensorName) {
        throw new Error('cameraId, sensorType, pluginId and sensorName are required for sensor storage removal');
      }
      storageKey = sensorStorageKey(deviceIdOrCameraId, sensorTypeOrUndefined, pluginId, sensorName);
    } else if (type === 'camera') {
      if (!deviceIdOrCameraId) {
        throw new Error('deviceId is required for camera storage removal');
      }
      storageKey = deviceIdOrCameraId;
    } else {
      storageKey = 'storage';
    }

    const deviceStorage = this.#storages.get(storageKey);
    await deviceStorage?.destroy();
    await deviceStorage?.unregisterStorage();
    this.#storages.delete(storageKey);
  }

  /** Internal method to close all storages */
  public async close(): Promise<void> {
    for (const storage of this.#storages.values()) {
      await storage.close();
    }
  }
}

File: server/src/plugins/runtime/python/storage_controller.py
from __future__ import annotations

from typing import TYPE_CHECKING, Literal, overload

from plugins.runtime.python.storage import DeviceStorage
from plugins.runtime.python.store import CameraLocation, PluginLocation, SensorLocation

if TYPE_CHECKING:
    from _camera_ui_tools.camera_ui_rpc import RPCClient
    from _camera_ui_tools.camera_ui_sdk import JsonSchema, SensorType
    from plugins.runtime.python.config_db import PluginConfigDb
    from plugins.runtime.python.plugin_api import PluginAPI
    from plugins.runtime.python.typings import PluginInfo


# Stable storage key for sensor instance storage.
def sensor_storage_key(camera_id: str, sensor_type: str, plugin_id: str, sensor_name: str) -> str:
    return f"{camera_id}:sensor:{sensor_type}:{plugin_id}:{sensor_name}"


class StorageController:
    def __init__(
        self,
        api: PluginAPI,
        proxy: RPCClient,
        plugin: PluginInfo,
        plugin_db: PluginConfigDb,
    ) -> None:
        self.__api: PluginAPI = api
        self.__proxy = proxy
        self.__plugin = plugin
        self.__storages: dict[str, DeviceStorage] = {}
        self.__plugin_db = plugin_db

    def createCameraStorage(self, cameraId: str, schemas: list[JsonSchema] | None = []) -> DeviceStorage:
        if schemas is None:
            schemas = []

        camera_storage = self.__storages.get(cameraId)

        if not camera_storage:
            camera_storage = DeviceStorage(
                self.__api,
                self.__proxy,
                self.__plugin,
                self.__plugin_db,
                CameraLocation(camera_id=cameraId),
                schemas,
            )
            self.__storages[cameraId] = camera_storage
        else:
            camera_storage.update_schema(schemas)

        return camera_storage

    def createPluginStorage(self, schemas: list[JsonSchema] | None = []) -> DeviceStorage:
        if schemas is None:
            schemas = []

        plugin_storage = self.__storages.get("storage")

        if not plugin_storage:
            plugin_storage = DeviceStorage(
                self.__api,
                self.__proxy,
                self.__plugin,
                self.__plugin_db,
                PluginLocation(),
                schemas,
            )
            self.__storages["storage"] = plugin_storage
        else:
            plugin_storage.update_schema(schemas)

        return plugin_storage

    def getCameraStorage(self, camera_id: str) -> DeviceStorage | None:
        return self.__storages.get(camera_id)

    def getPluginStorage(self) -> DeviceStorage | None:
        return self.__storages.get("storage")

    def createSensorStorage(
        self,
        camera_id: str,
        sensor_type: SensorType,
        plugin_id: str,
        sensor_name: str,
        sensor_id: str,
        schemas: list[JsonSchema] = [],
    ) -> DeviceStorage:
        storage_key = sensor_storage_key(camera_id, sensor_type.value, plugin_id, sensor_name)
        storage = self.__storages.get(storage_key)

        if not storage:
            # storage_key is only the in-memory registry key; persistence addresses
            # the canonical sensors.<camId>.<type>.<name> path. sensor_id is the
            # runtime UUID for the RPC namespace.
            storage = DeviceStorage(
                self.__api,
                self.__proxy,
                self.__plugin,
                self.__plugin_db,
                SensorLocation(camera_id=camera_id, sensor_type=sensor_type.value, sensor_name=sensor_name),
                schemas,
                sensor_id,
            )
            self.__storages[storage_key] = storage
            storage.update_schema(schemas)
        else:
            storage.update_schema(schemas)

        return storage

    def getSensorStorage(
        self, camera_id: str, sensor_type: SensorType, plugin_id: str, sensor_name: str
    ) -> DeviceStorage | None:
        storage_key = sensor_storage_key(camera_id, sensor_type.value, plugin_id, sensor_name)
        return self.__storages.get(storage_key)

    @overload
    async def createStorage(self, type_: Literal["camera"], device_id: str) -> DeviceStorage: ...
    @overload
    async def createStorage(self, type_: Literal["plugin"], device_id: None = None) -> DeviceStorage: ...
    @overload
    async def createStorage(
        self,
        type_: Literal["sensor"],
        device_id: str,
        sensor_type: SensorType,
        plugin_id: str,
        sensor_name: str,
        sensor_id: str,
    ) -> DeviceStorage: ...
    async def createStorage(
        self,
        type_: Literal["camera", "plugin", "sensor"],
        device_id: str | None = None,  # camera_id
        sensor_type: SensorType | None = None,
        plugin_id: str | None = None,
        sensor_name: str | None = None,
        sensor_id: str | None = None,
    ) -> DeviceStorage:
        storage: DeviceStorage | None = None

        if type_ == "camera":
            if not device_id:
                raise ValueError("ID is required for storage creation")

            storage = self.createCameraStorage(device_id)
        elif type_ == "sensor":
            if not device_id or not sensor_type or not plugin_id or not sensor_name or not sensor_id:
                raise ValueError(
                    "cameraId, sensorType, pluginId, sensorName and sensorId are required for sensor storage creation"
                )

            storage = self.createSensorStorage(device_id, sensor_type, plugin_id, sensor_name, sensor_id)
        else:
            storage = self.createPluginStorage()

        await storage.register_storage()
        return storage

    @overload
    async def removeStorage(self, type_: Literal["camera"], device_id: str) -> None: ...
    @overload
    async def removeStorage(self, type_: Literal["plugin"], device_id: None) -> None: ...
    @overload
    async def removeStorage(
        self,
        type_: Literal["sensor"],
        device_id: str | None = None,
        sensor_type: SensorType | None = None,
        plugin_id: str | None = None,
        sensor_name: str | None = None,
    ) -> None: ...
    async def removeStorage(
        self,
        type_: Literal["camera", "plugin", "sensor"],
        device_id: str | None = None,
        sensor_type: SensorType | None = None,
        plugin_id: str | None = None,
        sensor_name: str | None = None,
    ) -> None:
        storage_key: str

        if type_ == "sensor":
            if not device_id or not sensor_type or not plugin_id or not sensor_name:
                raise ValueError(
                    "cameraId, sensorType, pluginId and sensorName are required for sensor storage removal"
                )

            storage_key = sensor_storage_key(device_id, sensor_type.value, plugin_id, sensor_name)
        elif type_ == "camera":
            if not device_id:
                raise ValueError("ID is required for storage removal")
            storage_key = device_id
        else:
            storage_key = "storage"

        device_storage = self.__storages.get(storage_key)
        if device_storage:
            await device_storage.destroy()
            await device_storage.unregister_storage()
            del self.__storages[storage_key]
            return

    # Internal method to close all storages
    async def close(self) -> None:
        for storage in self.__storages.values():
            await storage.close()
